File: src/web/web_app/for_reference/legacy_desktop/main_window/main_widget/learn_tab/lesson_widget/question_generator.py
from __future__ import annotations
import random
from typing import TYPE_CHECKING

from enums.letter.letter import Letter
from main_window.main_widget.grid_mode_checker import GridModeChecker

from data.constants import END_POS, LETTER, START_POS

if TYPE_CHECKING:
    from main_window.main_widget.learn_tab.lesson_widget.lesson_widget import (
        LessonWidget,
    )


class QuestionGenerator:
    """
    A single unified question generator that dynamically generates different types of questions
    based on lesson configuration.
    """

    def __init__(self, lesson_widget: "LessonWidget", quiz_description: str) -> None:
        super().__init__()
        self.lesson_widget = lesson_widget
        self.main_widget = lesson_widget.main_widget
        self.quiz_description = quiz_description
        self.previous_correct_letter = None
        self.previous_pictographs = set()

    def generate_question(self):
        """
        Dynamically determines how to generate a question based on question type.
        """
        self.lesson_widget.update_progress_label()

        if self.quiz_description == "pictograph_to_letter":
            self._generate_pictograph_to_letter()
        elif self.quiz_description == "letter_to_pictograph":
            self._generate_letter_to_pictograph()
        elif self.quiz_description == "valid_next_pictograph":
            self._generate_valid_next_pictograph()
        else:
            raise ValueError(f"Unknown question type: {self.quiz_description}")

    def _generate_pictograph_to_letter(self):
        """Generates a question where a pictograph is displayed, and the answer is a letter."""
        correct_answer = self._generate_correct_letter()
        self.previous_correct_letter = correct_answer
        pictograph_dataset = self.main_widget.pictograph_dataset

        correct_pictograph_data = random.choice(pictograph_dataset[correct_answer])
        self.lesson_widget.question_widget.renderer.update_question(
            correct_pictograph_data
        )

        wrong_answers = self._generate_wrong_letters(correct_answer)
        options = [correct_answer.value] + wrong_answers
        random.shuffle(options)

        self.lesson_widget.answers_widget.update_answer_options(
            options,
            correct_answer.value,
            self.lesson_widget.answer_checker.check_answer,
        )

    def _generate_letter_to_pictograph(self):
        """Generates a question where a letter is displayed, and the answer is a pictograph."""
        correct_letter, correct_pictograph = self._generate_correct_pictograph()

        self.lesson_widget.question_widget.renderer.update_question(
            correct_letter.value
        )

        pictographs = self._get_shuffled_pictographs(correct_pictograph)
        self.lesson_widget.answers_widget.update_answer_options(
            pictographs,
            correct_pictograph,
            self.lesson_widget.answer_checker.check_answer,
        )

    def _get_shuffled_pictographs(self, correct_pictograph):
        """Generate and shuffle the correct and wrong pictographs."""
        wrong_pictographs = self.generate_wrong_pictographs(correct_pictograph[LETTER])
        pictographs = [correct_pictograph] + wrong_pictographs
        random.shuffle(pictographs)

        return pictographs

    def _get_available_letters(self, correct_letter: str) -> list[str]:
        """Get the available letters excluding the correct letter."""
        return [
            letter
            for letter in self.main_widget.pictograph_dataset
            if letter != correct_letter
        ]

    def generate_wrong_pictographs(self, correct_letter: str) -> list[dict]:
        """Generate three random wrong pictographs, ensuring each has a different letter."""
        available_letters = self._get_available_letters(correct_letter)
        wrong_pictographs = []

        while len(wrong_pictographs) < 3:
            # Ensure the letter hasn't been used before and does not match the correct letter
            letter, pictograph_data = self._get_random_pictograph(available_letters)

            if letter.value != correct_letter:
                wrong_pictographs.append(pictograph_data)
                available_letters.remove(letter)

        return wrong_pictographs

    def _generate_valid_next_pictograph(self):
        """Generates a question where a pictograph is shown, and the answer is the valid next pictograph."""
        initial_pictograph = self._generate_initial_pictograph()
        self.previous_pictograph = initial_pictograph

        self.lesson_widget.question_widget.renderer.update_question(initial_pictograph)

        correct_pictograph = self._generate_correct_next_pictograph(initial_pictograph)
        wrong_pictographs = self._generate_wrong_pictographs(correct_pictograph)

        options = [correct_pictograph] + wrong_pictographs
        random.shuffle(options)

        self.lesson_widget.answers_widget.update_answer_options(
            options, correct_pictograph, self.lesson_widget.answer_checker.check_answer
        )

    def _generate_correct_letter(self) -> Letter:
        """Generates a correct letter, ensuring it is not the same as the previous one."""
        letters = list(self.main_widget.pictograph_dataset.keys())
        if self.previous_correct_letter:
            letters.remove(self.previous_correct_letter)
        return random.choice(letters)

    def _generate_correct_pictograph(self) -> tuple[Letter, dict]:
        """Retrieves a correct letter and its corresponding pictograph."""
        available_letters = list(self.main_widget.pictograph_dataset.keys())
        correct_letter = random.choice(available_letters)
        pictograph_datas = self.filter_pictograph_dataset_by_grid_mode()
        correct_pictograph = random.choice(pictograph_datas[correct_letter])
        return correct_letter, correct_pictograph

    def _generate_wrong_letters(self, correct_letter: Letter) -> list[str]:
        """Generates three incorrect letter answers."""
        return random.sample(
            [
                letter.value
                for letter in self.main_widget.pictograph_dataset
                if letter != correct_letter
            ],
            3,
        )

    def _generate_wrong_pictographs(self, correct_pictograph: dict) -> list[dict]:
        """Generates three incorrect pictographs."""
        wrong_pictographs = []
        while len(wrong_pictographs) < 3:
            random_pictograph = self._generate_random_pictograph()
            if random_pictograph[START_POS] != correct_pictograph[START_POS]:
                wrong_pictographs.append(random_pictograph)
        return wrong_pictographs

    def _generate_initial_pictograph(self) -> dict:
        """Generates an initial pictograph to display."""
        available_letters = list(self.main_widget.pictograph_dataset.keys())
        pictograph_datas = self.filter_pictograph_dataset_by_grid_mode()

        available_letters = [
            letter
            for letter in available_letters
            for pictograph in pictograph_datas[letter]
            if pictograph[START_POS] == pictograph[END_POS]
        ]
        letter = random.choice(available_letters)
        return random.choice(pictograph_datas[letter])

    def _generate_correct_next_pictograph(self, initial_pictograph: dict) -> dict:
        """Finds a valid pictograph that can follow the initial pictograph."""
        end_pos = initial_pictograph[END_POS]
        pictograph_datas = self.filter_pictograph_dataset_by_grid_mode()
        valid_pictographs = [
            pictograph
            for letter_pictographs in pictograph_datas.values()
            for pictograph in letter_pictographs
            if pictograph[START_POS] == end_pos
        ]

        correct_answer = random.choice(valid_pictographs)
        return correct_answer

    def _generate_random_pictograph(self) -> dict:
        """Selects a random pictograph."""
        pictograph_datas = self.filter_pictograph_dataset_by_grid_mode()
        while True:
            letter = random.choice(list(self.main_widget.pictograph_dataset.keys()))
            random_pictograph = random.choice(pictograph_datas[letter])
            if random_pictograph[START_POS] != random_pictograph[END_POS]:
                return random_pictograph

    def _get_random_pictograph(
        self, available_letters: list[str]
    ) -> tuple[Letter, dict]:
        """Choose a random letter and a corresponding pictograph."""
        letter = random.choice(available_letters)
        pictograph_datas = self.filter_pictograph_dataset_by_grid_mode()
        pictograph_data = random.choice(pictograph_datas[letter])
        return letter, pictograph_data

    def filter_pictograph_dataset_by_grid_mode(self) -> dict[Letter, list[dict]]:
        """Filter pictograph dicts by grid mode."""
        valid_dicts: dict[Letter, list[dict]] = {}
        for letter in self.main_widget.pictograph_dataset:
            valid_dicts.setdefault(letter, [])
            for pictograph_data in self.main_widget.pictograph_dataset[letter]:
                if (
                    GridModeChecker.get_grid_mode(pictograph_data)
                    # == grid_mode
                ):
                    valid_dicts[letter].append(pictograph_data)
        return valid_dicts

    def fade_to_new_question(self):
        widgets_to_fade = [
            self.lesson_widget.question_widget,
            self.lesson_widget.answers_widget,
        ]
        self.lesson_widget.fade_manager.widget_fader.fade_and_update(
            widgets_to_fade,
            callback=self.generate_question,
        )

File: src/lib/modules/learn/quiz/services/implementations/QuestionGenerator.ts
/**
 * Question Generator Service - 2025 Modern Implementation
 *
 * Generates quiz questions for all three lesson types using real pictograph data.
 * Implements the logic from the legacy desktop app with modern TypeScript patterns.
 */

import type { Letter } from "$lib/shared/foundation/domain/models/Letter";
import type { PictographData } from "$lib/shared/pictograph/shared/domain/models/PictographData";
import { resolve } from "$lib/shared/inversify";
import { TYPES } from "$lib/shared/inversify/types";
import type { ILetterQueryHandler } from "$lib/shared/foundation/services/contracts/data";
import { GridMode } from "$lib/shared/pictograph/grid/domain/enums/grid-enums";
import {
  QuizAnswerFormat,
  QuizQuestionFormat,
  QuizType,
  type QuizAnswerOption,
  type QuizQuestionData,
} from "../../domain";

export class QuestionGeneratorService {
  private static letterQueryHandler: ILetterQueryHandler | null = null;
  private static previousCorrectLetter: Letter | null = null;
  private static allPictographs: PictographData[] = [];
  private static pictographsByLetter: Map<Letter, PictographData[]> = new Map();
  private static availableLetters: Letter[] = [];
  private static isInitialized = false;

  /**
   * Initialize the service - must be called before generating questions
   * Loads pictographs from static CSV files using ILetterQueryHandler
   */
  static async initialize(): Promise<void> {
    if (this.isInitialized) return;

    try {
      console.log("🔄 QuestionGenerator: Initializing ILetterQueryHandler...");
      this.letterQueryHandler = resolve<ILetterQueryHandler>(
        TYPES.ILetterQueryHandler
      );

      // Load ALL pictograph variations from CSV (Diamond mode)
      console.log("🔄 QuestionGenerator: Loading pictographs from CSV...");
      this.allPictographs =
        await this.letterQueryHandler.getAllPictographVariations(
          GridMode.DIAMOND
        );
      console.log(
        `✅ QuestionGenerator: Loaded ${this.allPictographs.length} pictographs`
      );

      // Log first pictograph to see its structure
      if (this.allPictographs.length > 0) {
        console.log("📝 QuestionGenerator: Sample pictograph:", {
          letter: this.allPictographs[0]!.letter,
          startPosition: this.allPictographs[0]!.startPosition,
          endPosition: this.allPictographs[0]!.endPosition,
          hasMotions: !!this.allPictographs[0]!.motions,
        });
      }

      // Group by letter
      this.pictographsByLetter.clear();
      let pictographsWithLetters = 0;
      this.allPictographs.forEach((picto) => {
        if (picto.letter) {
          pictographsWithLetters++;
          const existing = this.pictographsByLetter.get(picto.letter) || [];
          existing.push(picto);
          this.pictographsByLetter.set(picto.letter, existing);
        }
      });

      console.log(
        `📊 QuestionGenerator: ${pictographsWithLetters} pictographs have letters`
      );

      // Get available letters
      this.availableLetters = Array.from(this.pictographsByLetter.keys());
      console.log(
        `✅ QuestionGenerator: ${this.availableLetters.length} letters available:`,
        this.availableLetters.join(", ")
      );

      this.isInitialized = true;
    } catch (error) {
      console.error("❌ QuestionGenerator: Failed to initialize:", error);
      throw error;
    }
  }

  /**
   * Generate a question for a specific quiz type
   */
  static async generateQuestion(quizType: QuizType): Promise<QuizQuestionData> {
    if (!this.isInitialized) {
      await this.initialize();
    }

    if (this.availableLetters.length === 0) {
      const errorMsg = `No pictographs available to generate questions. Loaded: ${this.allPictographs.length} total pictographs, ${this.pictographsByLetter.size} have letters. Initialized: ${this.isInitialized}`;
      console.error("❌ QuestionGenerator:", errorMsg);
      throw new Error(errorMsg);
    }

    const questionId = this.generateQuestionId();

    switch (quizType) {
      case QuizType.PICTOGRAPH_TO_LETTER:
        return this.generatePictographToLetterQuestion(questionId);
      case QuizType.LETTER_TO_PICTOGRAPH:
        return this.generateLetterToPictographQuestion(questionId);
      case QuizType.VALID_NEXT_PICTOGRAPH:
        return this.generateValidNextPictographQuestion(questionId);
      default:
        throw new Error(`Unsupported quiz type: ${quizType}`);
    }
  }

  /**
   * LESSON 1: Pictograph to Letter
   * Show a pictograph, user picks the correct letter
   */
  private static generatePictographToLetterQuestion(
    questionId: string
  ): QuizQuestionData {
    // Pick a random letter (avoid repeating the same letter)
    const correctLetter = this.getRandomLetter();
    const correctPictographs = this.pictographsByLetter.get(correctLetter);

    if (!correctPictographs || correctPictographs.length === 0) {
      throw new Error(`No pictographs found for letter: ${correctLetter}`);
    }

    // Pick a random pictograph for this letter
    const correctPictograph = this.getRandomItem(correctPictographs);

    // Generate 3 wrong letters
    const wrongLetters = this.generateWrongLetters(correctLetter, 3);

    // Create answer options (correct + 3 wrong)
    const allLetters = [correctLetter, ...wrongLetters];
    this.shuffleArray(allLetters);

    const answerOptions: QuizAnswerOption[] = allLetters.map((letter) => ({
      id: this.generateOptionId(),
      content: letter,
      isCorrect: letter === correctLetter,
    }));

    return {
      questionId,
      questionContent: correctPictograph,
      answerOptions,
      correctAnswer: correctLetter,
      questionType: QuizQuestionFormat.PICTOGRAPH,
      answerType: QuizAnswerFormat.BUTTON,
      lessonType: QuizType.PICTOGRAPH_TO_LETTER,
      generationTimestamp: new Date().toISOString(),
    };
  }

  /**
   * LESSON 2: Letter to Pictograph
   * Show a letter, user picks the correct pictograph
   */
  private static generateLetterToPictographQuestion(
    questionId: string
  ): QuizQuestionData {
    // Pick a random letter
    const correctLetter = this.getRandomLetter();
    const correctPictographs = this.pictographsByLetter.get(correctLetter);

    if (!correctPictographs || correctPictographs.length === 0) {
      throw new Error(`No pictographs found for letter: ${correctLetter}`);
    }

    // Pick a random pictograph for this letter
    const correctPictograph = this.getRandomItem(correctPictographs);

    // Generate 3 wrong pictographs (from different letters)
    const wrongPictographs = this.generateWrongPictographs(correctLetter, 3);

    // Create answer options
    const allPictographs = [correctPictograph, ...wrongPictographs];
    this.shuffleArray(allPictographs);

    const answerOptions: QuizAnswerOption[] = allPictographs.map(
      (pictograph) => ({
        id: this.generateOptionId(),
        content: pictograph,
        isCorrect: pictograph.letter === correctLetter,
      })
    );

    return {
      questionId,
      questionContent: correctLetter,
      answerOptions,
      correctAnswer: correctPictograph,
      questionType: QuizQuestionFormat.LETTER,
      answerType: QuizAnswerFormat.PICTOGRAPH,
      lessonType: QuizType.LETTER_TO_PICTOGRAPH,
      generationTimestamp: new Date().toISOString(),
    };
  }

  /**
   * LESSON 3: Valid Next Pictograph
   * Show a pictograph, user picks which pictograph can follow it
   * Rule: Next pictograph's START_POS must equal initial pictograph's END_POS
   */
  private static generateValidNextPictographQuestion(
    questionId: string
  ): QuizQuestionData {
    // Generate initial pictograph (must have START_POS == END_POS)
    const initialPictograph = this.generateInitialPictograph();

    if (!initialPictograph.endPosition) {
      throw new Error("Initial pictograph missing endPosition");
    }

    // Find valid next pictograph (START_POS == initial's END_POS)
    const correctNextPictograph =
      this.generateCorrectNextPictograph(initialPictograph);

    // Generate 3 wrong next pictographs (START_POS != initial's END_POS)
    const wrongNextPictographs = this.generateWrongNextPictographs(
      initialPictograph,
      3
    );

    // Create answer options
    const allOptions = [correctNextPictograph, ...wrongNextPictographs];
    this.shuffleArray(allOptions);

    const answerOptions: QuizAnswerOption[] = allOptions.map((pictograph) => ({
      id: this.generateOptionId(),
      content: pictograph,
      isCorrect: pictograph.startPosition === initialPictograph.endPosition,
    }));

    return {
      questionId,
      questionContent: initialPictograph,
      answerOptions,
      correctAnswer: correctNextPictograph,
      questionType: QuizQuestionFormat.PICTOGRAPH,
      answerType: QuizAnswerFormat.PICTOGRAPH,
      lessonType: QuizType.VALID_NEXT_PICTOGRAPH,
      generationTimestamp: new Date().toISOString(),
    };
  }

  // ============================================================================
  // HELPER METHODS
  // ============================================================================

  /**
   * Get a random letter, avoiding the previous one
   */
  private static getRandomLetter(): Letter {
    let candidates = [...this.availableLetters];

    // Avoid repeating the same letter
    if (this.previousCorrectLetter && candidates.length > 1) {
      candidates = candidates.filter((l) => l !== this.previousCorrectLetter);
    }

    const letter = this.getRandomItem(candidates);
    this.previousCorrectLetter = letter;
    return letter;
  }

  /**
   * Generate wrong letters (different from correct)
   */
  private static generateWrongLetters(
    correctLetter: Letter,
    count: number
  ): Letter[] {
    const wrongLetters = this.availableLetters.filter(
      (letter) => letter !== correctLetter
    );
    return this.getRandomItems(wrongLetters, count);
  }

  /**
   * Generate wrong pictographs (from different letters)
   */
  private static generateWrongPictographs(
    correctLetter: Letter,
    count: number
  ): PictographData[] {
    const wrongLetters = this.generateWrongLetters(correctLetter, count);
    const wrongPictographs: PictographData[] = [];

    for (const letter of wrongLetters) {
      const pictographs = this.pictographsByLetter.get(letter);
      if (pictographs && pictographs.length > 0) {
        wrongPictographs.push(this.getRandomItem(pictographs));
      }
    }

    return wrongPictographs;
  }

  /**
   * Generate initial pictograph for Lesson 3
   * Must have startPosition === endPosition
   */
  private static generateInitialPictograph(): PictographData {
    const validPictographs = this.allPictographs.filter(
      (p) =>
        p.startPosition && p.endPosition && p.startPosition === p.endPosition
    );

    if (validPictographs.length === 0) {
      // Fallback: just pick any pictograph
      console.warn(
        "⚠️ No pictographs with startPos === endPos found, using fallback"
      );
      return this.getRandomItem(this.allPictographs);
    }

    return this.getRandomItem(validPictographs);
  }

  /**
   * Generate correct next pictograph for Lesson 3
   * Must have startPosition === initialPictograph.endPosition
   */
  private static generateCorrectNextPictograph(
    initialPictograph: PictographData
  ): PictographData {
    const endPos = initialPictograph.endPosition;

    const validNextPictographs = this.allPictographs.filter(
      (p) => p.startPosition === endPos
    );

    if (validNextPictographs.length === 0) {
      throw new Error(
        `No valid next pictographs found for endPosition: ${endPos}`
      );
    }

    return this.getRandomItem(validNextPictographs);
  }

  /**
   * Generate wrong next pictographs for Lesson 3
   * Must have startPosition !== initialPictograph.endPosition
   */
  private static generateWrongNextPictographs(
    initialPictograph: PictographData,
    count: number
  ): PictographData[] {
    const endPos = initialPictograph.endPosition;

    const invalidNextPictographs = this.allPictographs.filter(
      (p) => p.startPosition !== endPos
    );

    if (invalidNextPictographs.length < count) {
      console.warn(
        `⚠️ Only ${invalidNextPictographs.length} invalid next pictographs available, requested ${count}`
      );
      return this.getRandomItems(
        invalidNextPictographs,
        invalidNextPictographs.length
      );
    }

    return this.getRandomItems(invalidNextPictographs, count);
  }

  // ============================================================================
  // UTILITY METHODS
  // ============================================================================

  private static getRandomItem<T>(array: T[]): T {
    if (array.length === 0) {
      throw new Error("Cannot get random item from empty array");
    }
    return array[Math.floor(Math.random() * array.length)]!;
  }

  private static getRandomItems<T>(array: T[], count: number): T[] {
    const shuffled = [...array].sort(() => Math.random() - 0.5);
    return shuffled.slice(0, Math.min(count, array.length));
  }

  private static shuffleArray<T>(array: T[]): void {
    for (let i = array.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [array[i]!, array[j]!] = [array[j]!, array[i]!];
    }
  }

  private static generateQuestionId(): string {
    return `q_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
  }

  private static generateOptionId(): string {
    return `opt_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
  }

  /**
   * Reset generator state
   */
  static resetState(): void {
    this.previousCorrectLetter = null;
  }
}
